File: Sources/Networking/ExecuteRequests/Execute.repostWithCounters.swift
import Foundation
import SmithereenAPI

struct RepostWithCounters: Codable {
    var postID: WallPostID
    var repostsCount: Int
    var likesCount: Int

    private enum CodingKeys: String, CodingKey {
        case postID = "post_id"
        case repostsCount = "reposts_count"
        case likesCount = "likes_count"
    }
}

extension Execute<Wall.Repost, RepostWithCounters> {
    private static let script = scriptResource("repostWithCounters")

    static func repostWithCounters(
        postID: WallPostID,
        message: String? = nil,
        textFormat: TextFormat? = nil,
        attachments: [AttachmentToCreate]? = nil,
        contentWarning: String? = nil,
        guid: UUID? = nil,
    ) -> Self {
        Execute(
            code: script,
            args: Wall.Repost(
                postID: postID,
                message: message,
                textFormat: textFormat,
                attachments: attachments,
                contentWarning: contentWarning,
                guid: guid,
            )
        )
    }
}

File: Sources/Networking/ExecuteRequests/execute.repostWithCounters.js
var repostID = API.wall.repost(
	{
		post_id: Args.post_id,
		message: Args.message,
		text_format: Args.text_format,
		attachments: Args.attachments,
		content_warning: Args.content_warning,
		guid: Args.guid,
	}
);
var op = API.wall.getById(
	{
		posts: Args.post_id,
		repost_history_depth: 1,
	}
)[0];
if (op.is_mastodon_style_repost) {
  op = op.repost_history[0];
}
var reposts_count = 0;
if (op.reposts) {
  reposts_count = op.reposts.count;
}
var likes_count = 0;
if (op.likes) {
  likes_count = op.likes.count;
}
return {
	post_id: repostID,
	reposts_count: reposts_count,
	likes_count: likes_count,
};
